ctarguments.dd

```diff
@@ -0,0 +1,217 @@
+alias Numbers = Arguments!(1, 2, 3, 4);
+static assert (Numbers[1] == 2);
+alias SubNumbers = Numbers[1 .. $];
+static assert (SubNumbers[0] == 2);
+---
+
+$(H4 Assignment)
+
+$(P Works only if list element is a symbol that refers to a mutable variable)
+
+---
+import std.meta.arglist;
+
+void main()
+{
+    int x;
+    alias List = Arguments!(10, x);
+    List[1] = 42;
+    assert (x == 42);
+    // List[0] = 42; // won't compile, can't assign to a constant
+}
+---
+
+$(H4 Loops)
+
+$(P D $(LINK2 dlang.org/statement.html#ForeachStatement, foreach statement) has special
+    semantics when iterating over compile-time list. It duplicates body of the loop
+    for each of the list elements, with loop iteration "variable" becoming an alias
+    for that compile-time list elements.
+ )
+
+---
+import std.meta.arglist;
+
+void main()
+{
+    foreach (sym; Arguments!(int, "literal", main))
+    {
+        static if (is(sym))
+            pragma (msg, sym);
+        else
+            pragma (msg, typeof(sym));
+    }
+}
+
+/* Prints:
+
+   int
+   string
+   void()
+*/
+---
+
+$(H3 Auto-expansion)
+
+$(P One of less obvious properties of compile-time argument lists is that when used
+    as an argument to function or template those are automatically treated as list
+    of comma-separated arguments instead:
+ )
+
+---
+import std.meta.arglist;
+
+template Print0(T...)
+{
+    pragma(msg, T[0]);
+}
+
+alias Dummy = Print0!(Arguments!(int, double));
+---
+
+$(P This will only print `int` during compilation because last line gets rewritten
+    as `alias Dummy = Print0!(int, double)`. If auto-expansion didn't happen,
+    `Arguments!(int, double)` would be printed instead. This is inherent part of
+    language semantics for variadic lists and thus also preserved by `Arguments`.
+ )
+
+$(H3 Homogenous lists)
+
+$(P `Arguments` that only consists from only type or expression (value) elements is
+    commonly called "type list" or "expression list" accordingly. Concept of
+    "symbol list" is rarely mentioned explicitly but fits the same pattern.
+ )
+
+$(P It is possible to use homogenous type lists in declarations:)
+
+---
+import std.meta.arglist;
+alias Params = Arguments!(int, double, string);
+void foo(Params); // void foo(int, double, string);
+---
+
+$(P D supports special variable declaration syntax where type list acts as a type:)
+
+---
+import std.meta.arglist;
+
+void foo()
+{
+    Arguments!(int, double, string) variables;
+    variables[0] = 42;
+    variables[1] = 42.0;
+    variables[2] = "just a normal variable";
+}
+
+/* Compiler will rewrite such declaration to something like this:
+
+   int __var1;
+   double __var2;
+   string __var3;
+   alias variables = Arguments!(__var1, __var2, __var3);
+*/
+---
+
+$(P This is also what happens when you declare variadic template function:)
+
+---
+void foo(T...)(T args)
+{
+    // 'args' here is a compile-time list of symbols that
+    // refer to underlying compiler-generated arguments
```

Comment:
"underlying compiler-generated arguments" --. "the underlying compiler-generated arguments"